sycl/include/sycl/atomic.hpp

```diff
@@ -400,6 +400,11 @@ T atomic_fetch_max(atomic<T, addressSpace> Object, T Operand,
                    memory_order MemoryOrder = memory_order::relaxed) {
   return Object.fetch_max(Operand, MemoryOrder);
 }
+} // namespace detail
+template <typename T, access::address_space addressSpace =
+                          access::address_space::global_space>
+using atomic __SYCL2020_DEPRECATED("use sycl::atomic_ref instead") =
+    detail::atomic<T, addressSpace>;
```

Comment:
This needs a comment on what issues we're avoiding by introducing this alias.

Reply:
Added comment explaining introduction of this alias.

Reply:
Why can't we switch accessor to use `atomic_ref` instead?

Reply:
`atomic` is deprecated in SYCL 2020 but is still part of the specification. The interfaces in `accessor` that use `atomic` are also part of the specification and are also deprecated, but because `atomic` is deprecated it will cause deprecation warnings as soon as `accessor` is instantiated, even if the user does not intend to use the interfaces that use `atomic`.

Reply:
```c++
#ifdef BE_THE_HEADER
#pragma clang system_header
#include <type_traits>
template<typename T>
struct [[deprecated]] atomic{};

template<int Dimensions, typename DataT>
class accessor{
  public:
  template <int  Dims = Dimensions,
  typename = std::enable_if_t<(Dims > 0)>>
  int operator[](int Index) const {
    return Dims; 
  }

  template <int  Dims = Dimensions>
  typename std::enable_if_t<(Dims == 0), atomic<DataT>>
  operator[](int Index) const {
    return atomic<DataT>();
  }
};
#else
#define BE_THE_HEADER
#include __FILE__
int main(){
  accessor<1, int> test1;
  auto return_value = test1[0];
}
#endif
```

Compiling it with clang++ -Wall main.cpp -o main produced a deprecated warning, even though template specialization that utilizes atomic was not used.

Reply:
Yeah, this is an issue with Clang, it's morally equivalent to one we saw earlier this week using concepts: https://godbolt.org/z/bWdW1345r

I don't think anyone has filed an issue for it yet, but specializations that are instantiated during overload resolution seem to be diagnosing the attribute on them even if that overload is not selected. AIUI, this could be a bit of a challenge to fix in Clang, however.

Reply:
@DYNIO-INTEL , we need to create an issue upstream for our use case then.

@AaronBallman , do you know if there is workaround using some pragmas/attributes that won't force us to re-write our code like done in this PR?

Reply:
You may be able to use `if constexpr` to work around the issue, but there are no pragmas or attributes that can help here (that I'm aware of).

Reply:
I think the biggest issue is this implicit conversion function as it requires type of `atomic` and its instantiated for diagnostic issues deprecated warning. Is there any way to avoid it? 
```c++
template <int Dims = Dimensions,
            typename = std::enable_if_t<Dims == 0 &&
                                        AccessMode == access::mode::atomic>>
  operator
#ifdef __ENABLE_USM_ADDR_SPACE__
      atomic<DataT>
#else
      atomic<DataT, AS>
#endif
      () const {
    const size_t LinearIndex = getLinearIndex(id<AdjustedDim>());
    return Test(multi_ptr<DataT, AS, access::decorated::yes>(getQualifiedPtr() +
                                                             LinearIndex));
  }
```

Reply:
Can you provide a standalone godbolt link showing the issue?

Reply:
Sure, here it's https://godbolt.org/

Reply:
The only answer I have is pretty unsatisfying. I couldn't find a way to do this with `if constexpr` because it's really based on the declaration of the member function more than anything. The only thing I could think of is to not deprecate the class but to instead deprecate the interfaces which use the class: https://godbolt.org/z/WzY5cq86G
  
It's not ideal, but trying to change the behavior to not diagnose the attribute until after overload resolution would be a pretty big undertaking (as I understand it), without a clear path towards how to do it.